DSA_Session5/src/dictionary/DictionaryImp.java

```diff
@@ -0,0 +1,273 @@
+package dictionary;
+
+import java.util.*;
+
+import org.json.simple.JSONObject;
+
+/**
+ *Class DictionaryImp that implements dictionary interface
+ *
+ *@author Aman Gautam
+ * 
+ * Date - 06/08/2019
+ */
+public class DictionaryImp implements Dictionary {
+	BSTNode head;
+	
+	/**
+	 * parameterized constructor
+	 * @input json object
+   	 */
+	public DictionaryImp(JSONObject jsonObject) throws Exception {
+		this.head = null;
+		this.input(jsonObject);
+	}
+	
+	//default constructor
+	public DictionaryImp() {
+	}
+
+	/**
+	 * method for add key and value from json object
+	 * @param jsonObject
+	 */
+	private boolean input(JSONObject jsonObject) throws Exception{
+		
+		for(Object obj : jsonObject.keySet()){
+			String key = (String)obj;
+			String value = (String) jsonObject.get(key);
+			this.add(key, value);
+		}
+		return true;
+	}
+
+	/**
+	 * method for add key and value in BST
+	 */
+	@Override
+	public boolean add(String key, String value) throws Exception {
+		if (key.length() == 0 || value.length() == 0 || key == null || value == null) {
+			throw new NullPointerException("Invalid key and value");
+		}
+		if (head == null) {
+			head = new BSTNode(key, value);
+		} else {
+			addRec(key, value, head);
+		}
+		return false;
+	}
+
+	/**
+	 * Recursive function for adding key and value in BST
+	 * @param key
+	 * @param value
+	 * @param node
+	 * @return
+	 */
+	private boolean addRec(String key, String value, BSTNode node) {
+		// go left if key less than current node key
+		if (key.compareTo(node.getKey()) <= -1) {
+			if (node.getLeft() != null) {
+				// if left is present call recursive
+				addRec(key, value, node.getLeft());
+			} else {
+				// add value to left
+				node.setLeft(new BSTNode(key, value));
```

Comment:
Use the left and right variable to assign the left and right child, no need to use getLeft(), setLeft().